frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx

```diff
@@ -32,37 +36,88 @@ const resource = {
-  const [data, loaded, loadError] = useK8sWatchResource<CloudShellResource>(resource);
+  const [data, loaded, loadError] = useK8sWatchResource<CloudShellResource[]>(resource);
+  const [initData, setInitData] = React.useState<TerminalInitData>();
+  const [initDataLoading, setInitDataLoading] = React.useState<boolean>(false);
+  const [initError, setInitError] = React.useState<string>();
+  let workspace: CloudShellResource;
+  let workspaceName: string;
+  let workspaceNamespace: string;
+  let workspacePhase: string;
 
-  if (loadError) {
+  if (Array.isArray(data)) {
+    workspace = data.find(
+      (ws) => ws?.metadata?.annotations?.[CLOUD_SHELL_USER_ANNOTATION] === username,
+    );
+    workspacePhase = workspace?.status?.phase;
+    workspaceName = workspace?.metadata?.name;
+    workspaceNamespace = workspace?.metadata?.namespace;
+  }
+
+  React.useEffect(() => {
+    let unmounted = false;
+
+    if (workspacePhase === 'Running') {
+      setInitDataLoading(true);
+      initTerminal(username, workspaceName, workspaceNamespace)
+        .then((res: TerminalInitData) => {
+          if (unmounted) return;
+          setInitData(res);
+          setInitDataLoading(false);
+        })
+        .catch(() => {
+          if (unmounted) return;
+          setInitDataLoading(false);
+          setInitError('Failed to connect to your OpenShift command line terminal');
+        });
+    }
+
+    return () => {
+      unmounted = true;
+    };
+  }, [username, workspaceName, workspaceNamespace, workspacePhase]);
+
+  if (loadError || initError) {
     return (
-      <StatusBox loaded={loaded} loadError={loadError} label="OpenShift command line terminal" />
+      <StatusBox
+        loaded={loaded}
+        loadError={loadError || initError}
+        label="OpenShift command line terminal"
+      />
     );
   }
 
-  if (!loaded) {
-    return <LoadingBox />;
+  if (!loaded || initDataLoading) {
+    return (
+      <div className="odc-cloudshell-terminal__container">
+        <TerminalLoadingBox />
+      </div>
+    );
   }
 
-  if (Array.isArray(data)) {
-    const workspace = data.find(
-      (d) => d?.metadata?.annotations?.[CLOUD_SHELL_USER_ANNOTATION] === username,
+  if (initData && workspaceNamespace) {
+    return (
+      <div className="odc-cloudshell-terminal__container">
+        <CloudshellExec
+          namespace={workspaceNamespace}
+          container={initData.container}
+          podname={initData.pod}
+          shcommand={initData.cmd || []}
+        />
+      </div>
     );
-    if (workspace) {
-      const running = workspace.status?.phase === 'Running';
-      const url = workspace.status?.ideUrl;
-      return <CloudShellTerminalFrame loading={!running} url={url} />;
-    }
   }
 
-  return <CloudShellSetup onCancel={onCancel} />;
+  if (loaded && !workspace) return <CloudShellSetup onCancel={onCancel} />;
+
+  return null;
```

Comment:
I don't think we should ever return null. We have enough information to know if we should be showing loading, error, terminal or form at any point. The problem is with `initDataLoading` which I don't think is necessary. We can assume loading still if we do not have `initData`.